src/Craftwork/IdObfuscator/CharacterSet.php

```diff
@@ -0,0 +1,131 @@
+<?php
+declare(strict_types=1);
+
+namespace Craftwork\IdObfuscator;
+
+use Craftwork\IdObfuscator\Exception\DuplicateCharacterException;
+use Craftwork\IdObfuscator\Exception\InvalidCharacterSetException;
+use Craftwork\IdObfuscator\Exception\InvalidCharacterSetLengthException;
+
+final class CharacterSet
+{
+    const CHARSET_DEFAULT = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890-_';
+
+    /**
+     * @var string
+     */
+    private $characters;
+
+    /**
+     * @param string $characters
+     */
+    private function __construct(string $characters)
+    {
+        $this->assertAllAsciiCharacters($characters);
+        $this->assertCharactersLength($characters);
+        $this->assertNoDuplicateCharacters($characters);
+
+        $this->characters = $characters;
+    }
+
+    /**
+     * @param string $characters
+     *
+     * @return CharacterSet
+     *
+     * @throws InvalidCharacterSetLengthException
+     * @throws DuplicateCharacterException
+     * @throws InvalidCharacterSetException
+     */
+    public static function ofCustomCharacters(string $characters): CharacterSet
+    {
+        return new self($characters);
+    }
+
+    /**
+     * @return CharacterSet
+     */
+    public static function ofDefaultCharacters(): CharacterSet
+    {
+        return new self(self::CHARSET_DEFAULT);
+    }
+
+    /**
+     * @return string
+     */
+    public function getCharacters(): string
+    {
+        return $this->characters;
+    }
+
+    /**
+     * @param string $salt
+     * @return CharacterSet
+     */
+    public function shuffleCharacterSetWithSalt(string $salt): CharacterSet
```

Comment:
now that this method is in the CharacterSet class, what do you think about renaming it, e.g. CharacterSet::shuffleWithSalt?

Reply:
Actually, now I feel that the salt should be part of the character set class. A few things stick out at me:

* The shuffleCharacterSetWithSalt is public, but not really a necessary part of the contract
* In theory someone could reshuffle the character set by calling shuffleCharacterSetWithSalt outside of IdObfuscator resulting in broken/undesirable results
* Two character sets that have been shuffled with a different salt are not the same object at the end of the day

By moving the salt to the character set class, the interface/contract is cleaner and it becomes immutable because you can't alter the state of the object after construction.

Reply:
I do agree with these points, but I'm not sure I like complicating what should be a simple VO and what it will mean in terms of the method signatures we have, A CharacterSet should IMO be just that: a set of characters. I also think salting belongs in the encryption world, so may better suited in remaining in the same place as encoding/decoding.

If we start salting upon CharacterSet construction it will mean we have to start doing something like `new CharacterSet($chars, $salt);` which seems unintuitive. Also I think we'll still need a salt in the Obfuscator itself, otherwise the default CharacterSet we provide to users will be unsalted/salted the same everywhere.

I think there are issues either way and I'm sure it's something we can resolve in a nice way, but potentially it's worth raising a separate issue and fixing there, as I don't think it's worth holding this PR up for.

Reply:
I agree with @amneale. 

Regarding reshuffling outside IdObfuscator, as a new instance of CharacterSet is returned it shouldn't lead to undesirable results.

Regarding the contract instead, this class could be renamed to something like ShuffableCharacterSet (if that's even a word) so that it will describe better what its purpose is.

Reply:
Fair enough. I was thinking it should be called SaltedCharacterSet and that it takes ownership of the salting. But yeah let's look at this as a separate ticket, maybe I'll mess around afterwords and show some diffs to see if there's anything that people like.